integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-spark-service-bundle/nifi-spark-service/src/main/java/com/thinkbiganalytics/nifi/v2/spark/SparkJobserverService.java

```diff
@@ -0,0 +1,357 @@
+                        sparkContextsStates.remove(contextName);
+                        getLogger().info(String.format("Deleted context %s from Spark Jobserver %s", contextName, jobServerUrl));
+                    }
+                }
+            } catch (Exception ex) {
+                getLogger().error(ex.getMessage());
+            }
+        }
+        return contextDeleted;
+    }
+
+    @Override
+    public boolean executeSparkContextJob(String appName, String classPath, String contextName, String args, boolean async) {
+
+        String id = contextName + System.nanoTime();
+
+        synchronized (sparkContextsStates) {
+            if (sparkContextsStates.containsKey(contextName)) {
+                sparkContextsStates.get(contextName).addExecutionLock(id);
+            }
+        }
+
+        SparkJobResult jobResult = null;
+        Boolean success = false;
+
+        try {
+            getLogger().info(String.format("Executing Spark App %s %s on context %s with args %s on Spark Jobserver %s", appName, classPath, contextName, args, jobServerUrl));
+
+            Map<String, String> params = new HashMap<String, String>();
+
+            params.put(ISparkJobServerClientConstants.PARAM_APP_NAME, appName);
+            params.put(ISparkJobServerClientConstants.PARAM_CLASS_PATH, classPath);
+            params.put(ISparkJobServerClientConstants.PARAM_CONTEXT, contextName);
+
+            if (async) {
+                params.put(ISparkJobServerClientConstants.PARAM_SYNC, "false");
+            } else {
+                params.put(ISparkJobServerClientConstants.PARAM_SYNC, "true");
+                params.put(ISparkJobServerClientConstants.PARAM_TIMEOUT, syncTimeout);
+            }
+
+            jobResult = client.startJob(args, params);
+
+            getLogger().info(String.format("Executed %s %s on context %s on Spark Jobserver %s", appName, classPath, contextName, jobServerUrl));
+
+        } catch (Exception ex) {
+            getLogger().error(ex.getMessage());
+        }
+        finally {
+            synchronized (sparkContextsStates) {
+                if (sparkContextsStates.containsKey(contextName)) {
+                    sparkContextsStates.get(contextName).resetTimeoutTime();
+                    sparkContextsStates.get(contextName).removeExecutionLock(id);
+                }
+            }
+        }
+
+        if (jobResult != null && jobResult.getStatus() != "ERROR") {
+            success = true;
+        }
+        return success;
+    }
+
+    /**
+     * Using a timer task to periodically delete Spark contexts which have timed out
+     */
+    private void setupContextTimeoutTask() {
+
+        getLogger().info("Creating context timeout task");
+        TimerTask contextTimeoutTask = new TimerTask() {
+            @Override
+            public void run() {
+                synchronized (sparkContextsStates) {
+                    final long currentTime = System.nanoTime();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this useless assignment to local variable "currentTime". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1854)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this unused "currentTime" local variable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1481)